### ISSUE TITLE
Some extensions should be installed for the View function?

### ISSUE THREAD

**Issue by lcmgood** [Ikuyadeu/vscode-R#263]
The view() is a great function. The data frame or help document was not be displayed in the VSCode IDE when the view function was turned on. What kind of extensions should be installed in the VSCode or R. The error message as follow.

_> View(mtcars)
Error in .External2(C_dataviewer, x, title) : unable to start data viewer
In addition: Warning message:
In View(mtcars) : unable to open display_

I use VSCode ssh to my Ubuntu server (ver 16.04). The Microsoft R Open 3.5 was installed on the Ubuntu server.

**Comment by renkun-ken** [comment]
You may enable R session watcher to use `View()` inside VSCode.

Please see https://github.com/Ikuyadeu/vscode-R/wiki/R-Session-watcher.

**Comment by lcmgood** [comment]
Thank you for your  response. the View() was already turn on inside VSCode. Document can be display, but the data frame and list still don't work.  The error messages as follow.

_> View(mtcars)
Error in .External2(C_dataviewer, x, title) : unable to start data viewer
In addition: Warning message:
In View(mtcars) : unable to open display_

**Comment by renkun-ken** [comment]
Looks like you don't enable R session watcher since it replaces `base::View` which does not call `.External2(C_dataviewer, x, title)`. Did you restart or reload VSCode after you turn on `r.sessionWatcher`?

Would you like to post `ls.str(getOption("vscodeR"))`?

**Comment by lcmgood** [comment]
the response like this

_> ls.str(getOption("vscodeR"))
attach : function ()  
dataview : function (x, title)  
dataview_data_type : function (x)  
dataview_table : function (data)  
dir :  chr "/mnt/ssd1/Data/r-project/.vscode/vscode-R"
dir_plot_history :  chr "/tmp/RtmpUJKTND/images"
dir_session :  chr "/mnt/ssd1/Data/r-project/.vscode/vscode-R/34650"
globalenv_file :  chr "/mnt/ssd1/Data/r-project/.vscode/vscode-R/34650/globalenv.json"
new_plot : function ()  
pid :  int 34650
platform : List of 8
 $ OS.type   : chr "unix"
 $ file.sep  : chr "/"
 $ dynlib.ext: chr ".so"
 $ GUI       : chr "vscode"
 $ endian    : chr "little"
 $ pkgType   : chr "source"
 $ path.sep  : chr ":"
 $ r_arch    : chr ""
plot_file :  chr "/mnt/ssd1/Data/r-project/.vscode/vscode-R/34650/plot.png"
plot_history_file :  NULL
plot_updated :  logi FALSE
rebind : function (sym, value, ns)  
respond : function (command, ...)  
response_file :  chr "/mnt/ssd1/Data/r-project/.vscode/vscode-R/response.log"
tempdir :  chr "/tmp/RtmpUJKTND"
unbox : function (x)  
update : function (...)_

**Comment by renkun-ken** [comment]
Looks like R session watcher script is sourced. Would you like to type `View` and post its source code? Looks like it is not really replaced by our version of `View()`.

**Comment by lcmgood** [comment]
Thank you .

> View
_function (x, title) 
{
    check <- Sys.getenv("_R_CHECK_SCREEN_DEVICE_", "")
    msg <- "View() should not be used in examples etc"
    if (identical(check, "stop")) 
        stop(msg, domain = NA)
    else if (identical(check, "warn")) 
        warning(msg, immediate. = TRUE, noBreaks. = TRUE, domain = NA)
    if (missing(title)) 
        title <- paste("Data:", deparse(substitute(x))[1])
    as.num.or.char <- function(x) {
        if (is.character(x)) 
            x
        else if (is.numeric(x)) {
            storage.mode(x) <- "double"
            x
        }
        else as.character(x)
    }
    x0 <- as.data.frame(x)
    x <- as.list(format.data.frame(x0))
    rn <- row.names(x0)
    if (any(rn != seq_along(rn))) 
        x <- c(list(row.names = rn), x)
    if (!is.list(x) || !length(x) || !all(sapply(x, is.atomic)) || 
        !max(lengths(x))) 
        stop("invalid 'x' argument")
    if (grepl("darwin", R.version$os)) 
        check_for_XQuartz()
    invisible(.External2(C_dataviewer, x, title))
}
<bytecode: 0x1addda0>
<environment: namespace:utils>_

**Comment by renkun-ken** [comment]
Your `View` function is not the R session watcher version (https://github.com/Ikuyadeu/vscode-R/blob/master/R/init.R#L168).

Looks like some other scripts replaced `utils::View` after session watcher does. Would you like to post your `sessionInfo()` and `.Rprofile`?

**Comment by lcmgood** [comment]
**seesionInfo() as follow**

> sessionInfo()
_R version 3.5.3 (2019-03-11)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.6 LTS

Matrix products: default
BLAS: /opt/microsoft/ropen/3.5.3/lib64/R/lib/libRblas.so
LAPACK: /opt/microsoft/ropen/3.5.3/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] RevoUtils_11.0.3     RevoUtilsMath_11.0.0

loaded via a namespace (and not attached):
[1] compiler_3.5.3 jsonlite_1.6_  

**.Rprofile as follow**

_source(file.path(Sys.getenv(if (.Platform$OS.type == "windows") "USERPROFILE" else "HOME"), ".vscode-R", "init.R"))_

**Comment by renkun-ken** [comment]
I guess it is caused by `RevoUtils` or ` RevoUtilsMath`?

Would you try starting a fresh R session without these two packages attached? I guess your R distribution is MRO which automatically attach the packages and replace `View`?

**Comment by lcmgood** [comment]
Thank you. I try to detach `RevoUtils` and `RevoUtilsMath` use as follow 

```
pkg <- c("package:RevoUtils","package:RevoUtilsMath")
lapply(pkg, detach, character.only = TRUE, unload = TRUE)
```

then, I use sessionInfo() again

```
R version 3.5.3 (2019-03-11)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.6 LTS

Matrix products: default
BLAS: /opt/microsoft/ropen/3.5.3/lib64/R/lib/libRblas.so
LAPACK: /opt/microsoft/ropen/3.5.3/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.5.3 jsonlite_1.6 
```

You will not find `RevoUtils` and `RevoUtilsMath`, While the error is still there when I use `View()` to display dataframe.

```
 View(mtcars)
Error in .External2(C_dataviewer, x, title) : unable to start data viewer
In addition: Warning message:
In View(mtcars) : unable to open display
```

**Comment by renkun-ken** [comment]
If the packages replace `utils::View` on load or attach, then unloading them will be of no use. Is it possible that you start an R session that do not attach these packages at all?

If your R distribution has to attach these packages and replace `utils::View`, I'm afraid we can do nothing at the moment. Looks like the packages are attached after sourcing `.Rprofile`. You may try manually `source("~/.Rprofile")` again after R session started. vscode-R `init.R` will replace `utils::View` again.

**Comment by renkun-ken** [comment]
Sorry for not discovering that your `View` is exactly the original version rather than replaced by any script. It looks like https://github.com/Ikuyadeu/vscode-R/blob/master/R/init.R#L234 simply does not work in your case.

**Comment by renkun-ken** [comment]
Looks like `utils` is already attached before `init.R` is sourced so the rebinding of `utils::View` does not affect `View` even as `utils::View` is replaced.

**Comment by lcmgood** [comment]
I am confused, Thank you for you help.

**Comment by lcmgood** [comment]
I use `conflicts(detail=TRUE)` in R session and find this, is a clue to resolve this error?

```
> conflicts(detail=TRUE)
$`package:methods`
[1] "body<-"    "kronecker"

$`package:base`
[1] "body<-"    "kronecker"
```

**Comment by lcmgood** [comment]
Thank you. I find a way to resolve it. replace `View()` into `utils::View()` in Rscript. Then the data frame and list will be display.

**Comment by renkun-ken** [comment]
Yes, that's a nice walk-around.

If looks like your R session starts with `utils` attached so the R session watcher script could not replace `View` when utils is attached. Will find a way to fix this.

**Comment by renkun-ken** [comment]
Will fix `rebind` via 74b96bf so that if `utils` is attached before sourcing `init.R` it still works.
